From 4980c363af34fd1beee3558de6b668477167f7c9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Mon, 17 Nov 2025 17:20:49 +0100
Subject: [PATCH 2/2] boot: avoid increasing rollback counter during RAM load
 FW update

When MCUBOOT_RAM_LOAD_REVERT is enabled, the rollback counter must not
be incremented until image_ok is set.
Otherwise, in the event of a revert, the boot process may become bricked.

Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
---
 boot/bootutil/src/loader.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/boot/bootutil/src/loader.c b/boot/bootutil/src/loader.c
index bb6fa042..264bf76d 100644
--- a/boot/bootutil/src/loader.c
+++ b/boot/bootutil/src/loader.c
@@ -3158,10 +3158,12 @@ boot_update_hw_rollback_protection(struct boot_loader_state *state)
     /* Update the stored security counter with the newer (active) image's
      * security counter value.
      */
-#if defined(MCUBOOT_DIRECT_XIP) && defined(MCUBOOT_DIRECT_XIP_REVERT)
-    /* When the 'revert' mechanism is enabled in direct-xip mode, the
-     * security counter can be increased only after reboot, if the image
-     * has been confirmed at runtime (the image_ok flag has been set).
+#if (defined(MCUBOOT_DIRECT_XIP) && defined(MCUBOOT_DIRECT_XIP_REVERT)) || \
+    (defined(MCUBOOT_RAM_LOAD) && defined(MCUBOOT_RAM_LOAD_REVERT))
+    /* When the 'revert' mechanism is enabled in direct-xip or ram-load
+     * mode, the security counter can be increased only after reboot, if
+     * the image has been confirmed at runtime (the image_ok flag has
+     * been set).
      * This way a 'revert' can be performed when it's necessary.
      */
     if (state->slot_usage[BOOT_CURR_IMG(state)].swap_state.image_ok == BOOT_FLAG_SET) {
@@ -3173,7 +3175,8 @@ boot_update_hw_rollback_protection(struct boot_loader_state *state)
             BOOT_LOG_ERR("Security counter update failed after image %d validation.", BOOT_CURR_IMG(state));
             return rc;
         }
-#if defined(MCUBOOT_DIRECT_XIP) && defined(MCUBOOT_DIRECT_XIP_REVERT)
+#if (defined(MCUBOOT_DIRECT_XIP) && defined(MCUBOOT_DIRECT_XIP_REVERT)) || \
+    (defined(MCUBOOT_RAM_LOAD) && defined(MCUBOOT_RAM_LOAD_REVERT))
     }
 #endif
 
-- 
2.43.0

